### PR TITLE
bpo-34776: Fix dataclasses to support __future__ "annotations" mode

### PR DESCRIPTION
**Pull request by 1st1**
https://bugs.python.org/issue34776

### CODE REVIEW

**Review by gvanrossum — COMMENTED** [review]
I wonder if the fix shouldn't apply to all callers of `_create_fn()`? I know `__init__` is the only one that (currently) has annotations, but in general I wish the generated functions looked more like handwritten ones to various introspection tools.

**Comment by 1st1** [comment]
> I wonder if the fix shouldn't apply to all callers of _create_fn()? I know __init__ is the only one that (currently) has annotations, but in general I wish the generated functions looked more like handwritten ones to various introspection tools.

I think there's no harm in doing that.  I've updated the PR.  It's a bit more invasive now, but the benefit is that we *always* use the same mechanism (a closure) to inject objects into generated code.

**Comment by 1st1** [comment]
BTW this looks like a relatively safe change for 3.7.1, but it's Eric's call.

File: Lib/dataclasses.py [hunk 1]
```diff
@@ -347,20 +347,24 @@ def _create_fn(name, args, body, *, globals=None, locals=None,
     # __builtins__ may be the "builtins" module or
     # the value of its "__dict__",
     # so make sure "__builtins__" is the module.
-    if globals is not None and '__builtins__' not in globals:
-        globals['__builtins__'] = builtins
+    if '__builtins__' not in locals:
+        locals['__builtins__'] = builtins
```
Comment (gvanrossum) [inline Lib/dataclasses.py]: This doesn't actually work -- `exec()` gets `__builtins__` only from globals.

Comment (1st1) [reply Lib/dataclasses.py]: Hm, we only need  to bound the `"__builtins__"` name inside the closure as there's generated code that explicitly uses it as `"__builtins__.attribute"`. I can change the name to "_builtins" and everything should be fine too or... am I missing something here?

Comment (1st1) [reply Lib/dataclasses.py]: Probably it's better to rename it to BUILTINS, to make it more obvious.

Comment (1st1) [reply Lib/dataclasses.py]: Guido, I've updated the PR, please take a look.

File: Lib/dataclasses.py [hunk 2]
```diff
@@ -877,6 +893,7 @@ def _process_class(cls, init, repr, eq, order, unsafe_hash, frozen):
                                     # if possible.
                                     '__dataclass_self__' if 'self' in fields
                                             else 'self',
+                                    globals
```
Comment (gvanrossum) [inline Lib/dataclasses.py]: Trailing comma.

**Review by gvanrossum — COMMENTED** [review]
I don't follow the logic in dataclasses.py well enough to approve this, though I see nothing wrong other than a missing trailing comma. @ericvsmith will have to have the last word.

**Comment by 1st1** [comment]
According to https://discuss.python.org/t/3-7-1-and-3-6-7-release-update/184 there's still couple of days to get this into 3.7.1.  @ericvsmith do you have time for a review?

**Comment by 1st1** [comment]
FWIW we can just merge the 39006fb0ffbbf658a6d28e5c747cda64dc7520dd commit to 3.7 which does the trick in the most minimal way.

**Comment by ericvsmith** [comment]
Unfortunately I'm out of town until 2018-10-10 with limited internet access until then. I won't be able to do a decent review until then.

**Comment by 1st1** [comment]
Hi Eric, Python 3.7.2rc1 has been released but this PR is still not merged.  Is it possible to somehow unblock it?

**Comment by ericvsmith** [comment]
I apologize for being tied up with a client on end-of-year work. I don't see how this could get in to 3.7.2, but I'm planning on looking at it in January.

File: Misc/NEWS.d/next/Library/2018-09-23-14-24-37.bpo-34776.1SrQe3.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Fix dataclasses to support __future__ "annotations" mode
```
Comment (Vlad-Shcherbina) [inline Misc/NEWS.d/next/Library/2018-09-23-14-24-37.bpo-34776.1SrQe3.rst]: Strictly speaking, this problem is not directly related to `__future__` annotations.
It's about postponed evaluation of type annotations in general, including forward references in pre-PEP-563 annotations.
Consider rewording along the lines of "Fix dataclasses to support forward references in type annotations".

Here is an example that does not use PEP 563:

```python
from typing import get_type_hints
from dataclasses import dataclass

class T:
    pass

@dataclass()
class C2:
    x: 'T'

print(get_type_hints(C2.__init__))
```

Before your change:
```
Traceback (most recent call last):
  File ".\zzz.py", line 11, in <module>
    print(get_type_hints(C2.__init__))
  File "C:\Python37\lib\typing.py", line 1001, in get_type_hints
    value = _eval_type(value, globalns, localns)
  File "C:\Python37\lib\typing.py", line 260, in _eval_type
    return t._evaluate(globalns, localns)
  File "C:\Python37\lib\typing.py", line 464, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
NameError: name 'T' is not defined
```

After your change it works as expected:
```
{'x': <class '__main__.T'>, 'return': <class 'NoneType'>}
```

File: Lib/dataclasses.py [hunk 4]
```diff
@@ -368,23 +368,24 @@ def _create_fn(name, args, body, *, globals=None, locals=None,
...
-    body = '\n'.join(f' {b}' for b in body)
+    body = '\n'.join(f'  {b}' for b in body)
 
     # Compute the text of the entire function.
-    txt = f'def {name}({args}){return_annotation}:\n{body}'
+    txt = f' def {name}({args}){return_annotation}:\n{body}'
 
-    exec(txt, globals, locals)
-    return locals[name]
+    local_vars = ', '.join(locals.keys())
```
Comment (1st1) [inline Lib/dataclasses.py:383]: Note to self: add a comment explaining this.

**Comment by ilevkivskyi** [comment]
It looks like https://bugs.python.org/issue37948 will be also fixed by this PR.

**Review by ilevkivskyi — APPROVED** [review]
FWIW the fix looks good to me. It would be great to add few more comments and a test for plain string annotations and/or https://bugs.python.org/issue37948

**Comment by a-recknagel** [comment]
> It looks like https://bugs.python.org/issue37948 will be also fixed by this PR.

It won't, I just tested with https://github.com/1st1/cpython/tree/fix_dc_introspect

**Comment by ericvsmith** [comment]
> It won't, I just tested with https://github.com/1st1/cpython/tree/fix_dc_introspect

@a-recknagel : Could you post your test code and the results? Thanks.

**Comment by a-recknagel** [comment]
@ericvsmith  Sure. python was built from e38de430c92b6ea5d935910bfbae2744a469a595, and I ran

```python
>>> import dataclasses
>>> import typing
>>> A = dataclasses.make_dataclass('A', ['var'])
>>> typing.get_type_hints(A)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/arne/pythons/python_dc_fix/lib/python3.8/typing.py", line 972, in get_type_hints
    value = _eval_type(value, base_globals, localns)
  File "/home/arne/pythons/python_dc_fix/lib/python3.8/typing.py", line 259, in _eval_type
    return t._evaluate(globalns, localns)
  File "/home/arne/pythons/python_dc_fix/lib/python3.8/typing.py", line 463, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
NameError: name 'typing' is not defined
```
Which is the same behavior as on the current master.

**Pull request merged**